ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/query/BlockchainQueries.java

```diff
@@ -87,6 +96,24 @@ public BlockchainQueries(
             ? Optional.of(
                 new TransactionLogBloomCacher(blockchain, cachePath.get(), scheduler.get()))
             : Optional.empty();
+
+    blockBodySoftCache =
+        CacheBuilder.newBuilder()
+            .maximumSize(BLOCK_BODY_SOFT_CACHE_SIZE)
+            .softValues()
+            .build(CacheLoader.from(blockchain::getBlockBody));
+
+    blockchain.observeBlockAdded(this::onBlockAdded);
+    blockchain.observeChainReorg(this::onChainReorg);
+  }
+
+  private void onChainReorg(
+      final BlockWithReceipts blockWithReceipts, final Blockchain blockchain) {
+    blockBodySoftCache.invalidate(blockWithReceipts.getHash());
+  }
+
+  private void onBlockAdded(final BlockAddedEvent blockAddedEvent) {
+    blockBodySoftCache.invalidate(blockAddedEvent.getBlock().getHash());
```

Comment:
Why do we want to invalidate at all? Wouldn't a new block hash not exist in our cache?

Reply:
It will exist in the cache as an `Optional.empty()` - Guava's cache doesn't have a good way to handle a persistent cache miss, it models it that the result should then be a permanent "not found" value and if the underlying data changes we should invaldiate.

Reply:
Oh! Now I get it. We're invalidating the miss. Thank you